crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -922,6 +922,50 @@ impl<TX: DbTx, N: NodeTypes> StorageChangeSetReader for DatabaseProvider<TX, N>
     }
 }
 
+impl<TX: DbTx, N: NodeTypes> HashedStateReader for DatabaseProvider<TX, N> {
+    fn get_hashed_reverts(&self, from: BlockNumber) -> ProviderResult<HashedPostState> {
+        // Iterate over account changesets and record value before first occurring account change.
+        let mut accounts = HashMap::new();
+        let mut account_changesets_cursor = self.tx.cursor_read::<tables::AccountChangeSets>()?;
+        for entry in account_changesets_cursor.walk_range(from..)? {
+            let (_, AccountBeforeTx { address, info }) = entry?;
+            accounts.entry(address).or_insert(info);
+        }
+
+        // Iterate over storage changesets and record value before first occurring storage change.
+        let mut storages = HashMap::<Address, HashMap<B256, U256>>::default();
+        let mut storage_changesets_cursor = self.tx.cursor_read::<tables::StorageChangeSets>()?;
+        for entry in
+            storage_changesets_cursor.walk_range(BlockNumberAddress((from, Address::ZERO))..)?
+        {
+            let (BlockNumberAddress((_, address)), storage) = entry?;
+            let account_storage = storages.entry(address).or_default();
+            account_storage.entry(storage.key).or_insert(storage.value);
+        }
+
+        let hashed_accounts =
+            accounts.into_iter().map(|(address, info)| (keccak256(address), info)).collect();
```

Comment:
If we want to support generic key hashing then we should use the `KeyHasher` from the`StateCommitment` type. After #12895 is merged you can replace `keccak256` with `self.hash_key` (https://github.com/frisitano/reth/blob/776d374899b79f87ac2d80a954c02583aa22e08c/crates/storage/provider/src/providers/database/provider.rs#L386-L390).